src/components/Experiments/ExperimentEditModal.js

```diff
@@ -0,0 +1,195 @@
+import React, { Component } from 'react';
+import { connect } from 'react-redux';
+
+import {
+  Button,
+  Modal,
+  FormGroup,
+  FormControl,
+  Glyphicon,
+  Row,
+  Col,
+} from 'react-bootstrap';
+
+import {
+  showEditModal,
+  create as experimentCreate,
+  update as experimentUpdate,
+  uploadFilePairs,
+  UPLOAD_RES_SUCCESS,
+  UPLOAD_RES_FAILURE,
+} from '../../state/experiments';
+
+import './ExperimentEditModal.less';
+
+class ExperimentEditModal extends Component {
+  constructor(props) {
+    super(props);
+
+    const exp = this.props.editModalExperiment;
+    if (exp !== undefined) {
+      this.state = { nameVal: exp.name, descriptionVal: exp.description };
+    } else {
+      this.state = {
+        nameVal: '',
+        descriptionVal: '',
+      };
+    }
+
+    this.handleClose = this.handleClose.bind(this);
+    this.handleNameChange = this.handleNameChange.bind(this);
+    this.handleDescriptionChange = this.handleDescriptionChange.bind(this);
+    this.handleSave = this.handleSave.bind(this);
+    this.handleUpload = this.handleUpload.bind(this);
+  }
+
+  handleClose() {
+    this.props.showEditModal(false);
+  }
+
+  handleNameChange(e) {
+    this.setState({ nameVal: e.target.value });
+  }
+
+  handleDescriptionChange(e) {
+    this.setState({ descriptionVal: e.target.value });
+  }
+
+  handleSave() {
+    const create = this.props.editModalExperiment === undefined;
+
+    if (create) {
+      this.props.experimentCreate(
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    } else {
+      this.props.experimentUpdate(
+        this.props.editModalExperiment.id,
+        this.state.nameVal,
+        this.state.descriptionVal
+      );
+    }
+  }
+
+  handleUpload(e) {
+    this.props.uploadFilePairs(
+      this.props.editModalExperiment.id,
+      e.target.files[0]
+    );
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const exp = nextProps.editModalExperiment;
+    if (exp !== undefined) {
+      this.setState({ nameVal: exp.name, descriptionVal: exp.description });
+    } else {
+      this.setState({
+        nameVal: '',
+        descriptionVal: '',
+      });
+    }
+  }
+
+  render() {
+    const {
+      editModalShow,
+      createInProgress,
+      updateInProgress,
+      uploadInProgress,
+      uploadResult,
+    } = this.props.experiments;
+
+    const create = this.props.editModalExperiment === undefined;
+
+    const title = create ? 'Create Experiment' : 'Edit Experiment';
+    const btn = create ? 'Create' : 'Save';
+    const btnDisabled =
+      createInProgress || updateInProgress || this.state.nameVal.trim() === '';
+
+    const uploadDisabled = create || uploadInProgress;
+
+    let uploadIcon;
+    let uploadClass;
+    if (uploadInProgress) {
```

Comment:
looks like `switch` will be better fit here

Reply:
Are you sure? The chained if's condition use two different variables, I don't think it will be cleaner with a switch.

Reply:
and `if` &rarr; `else switch` would be clearer, but I'd not block this wonderful work for this cosmetic suggestion :wink: 
```js
if (uploadInProgress) { /* something */ }
else switch (uploadResult) {
    case UPLOAD_RES_SUCCESS: /* something */ break;
    case UPLOAD_RES_FAILURE: /* something */ break;
    default: /* something */
}
```